README.md
## Description

Yet another simple AviSynth plugin created for the sake of creating a simple AviSynth plugin.

This is a port of the great tophf's script [t_colormask][old_script]. It masks colors but unlike the old ColorKeyMask it works on YV12. The main reason for this plugin is performance. The API and output are not identical, please reffer to the **Difference** section for details.

### Requirements:

- AviSynth 2.60 / AviSynth+ 3.4 or later

- Microsoft VisualC++ Redistributable Package 2022 (can be downloaded from [here](https://github.com/abbodi1406/vcredist/releases))

### Usage:

```
tcolormask(clip, string "colors, int "tolerance", bool "bt601", bool "gray", int "lutthr", bool "mt", bool "onlyY", int "opt")
```

### Parameters:

- clip\
    A clip to process.\
    Must be in YUV420/422/444 8/16-bit planar format.

- colors\
    String of colors.\
    Line and block comments are allowed.

- tolerance\
    Pixel value will pass if its absolute difference with color is less than tolerance (luma) or half the tolerance (chroma).\
    It must be between 0 and 255/65535.\
    Default: 10(8-bit)/2570(16-bit).

- bt601\
    Use bt601 matrix for conversion of colors.\
    Default: False.

- gray\
    Set chroma of output clip to 128/32896.\
    Default: False.

- lutthr\
    If specified more than lutthr colors, lut will be used instead of direct SIMD computations.\
    Default: 9.

- mt\
    Enable multithreading.\
    It actually uses only two threads as any more threading doesn't seem to be useful.\
    Default: False.

- onlyY\
    The returned clip format is Y.\
    Default: False.

- opt\
    Sets which cpu optimizations to use.\
    -1: Auto-detect.\
    0: Use C++ code.\
    1: Use SSE2 code.\
    2: Use AVX2 code.\
    3: Use AVX512 code.\
    Default: -1.


### Example:

- 8-bit:

```
tcolormask("$FFFFFF /*pure white hex*/
000000 //also black
$808080", tolerance=10, bt601=false, gray=false, lutthr=9, mt=true)
```

- 16-bit:

```
tcolormask("$FFFFFFFFFFFF /*pure white hex*/
000000000000 //also black
$808080808080", tolerance=2570, bt601=false, gray=false, lutthr=9, mt=true)
```

### Difference:
Unlike the old script, this plugin uses a single string to specify all colors and doesn't do any blurring. Wrapper functions are provided for convenience in the **bin/tcolormask_wrappers.avs** script.

Also since we process chroma and luma together to avoid having to merge planes later, output is not [identical][comparison]. Basically it appears a bit less blurred and doesn't contain any non-binary values produced by chroma resizing. This will not be fixed and there's no workaround. You'll most likely be blurring the output clip anyway.

### Performance:
This plugin uses direct SIMD computations to do its dirty work. SIMD appears to be faster than LUT for the most common cases, but unfortunately its speed depends on the number of specified colors and at some point it does get slower than LUT. That's why the alternative is also provided. If specified more than *lutthr* (9 by default) colors, the plugin will use the LUT routine to avoid performance degradation.

All tests used YV12 1080p image cached by the *loop* function.

<table>
	<tr>
		<th>
			Colors
		</th>
		<th>
			Calls
		</th>
		<th>
			Old, fps
		</th>
		<th>
			SIMD ST, fps
		</th>
		<th>
			LUT ST, fps
		</th>
		<th>
			SIMD MT, fps
		</th>
		<th>
			LUT MT, fps
		</th>
	</tr>
	<tr>
		<td>
			1
		</td>
		<td>
			1
		</td>
		<td>
			52
		</td>
		<td>
			585
		</td>
		<td>
			150
		</td>
		<td>
			835
		</td>
		<td>
			265
		</td>
	</tr>
	<tr>
		<td>
			1
		</td>
		<td>
			10
		</td>
		<td>
			2
		</td>
		<td>
			82
		</td>
		<td>
			16
		</td>
		<td>
			122
		</td>
		<td>
			31
		</td>
	</tr>
	<tr>
		<td>
			6
		</td>
		<td>
			1
		</td>
		<td>
			52
		</td>
		<td>
			229
		</td>
		<td>
			150
		</td>
		<td>
			396
		</td>
		<td>
			265
		</td>
	</tr>
	<tr>
		<td>
			6
		</td>
		<td>
			10
		</td>
		<td>
			2
		</td>
		<td>
			26
		</td>
		<td>
			16
		</td>
		<td>
			47
		</td>
		<td>
			31
		</td>
	</tr>
	<tr>
		<td>
			24
		</td>
		<td>
			1
		</td>
		<td>
			52
		</td>
		<td>
			71
		</td>
		<td>
			150
		</td>
		<td>
			130
		</td>
		<td>
			265
		</td>
	</tr>
	<tr>
		<td>
			24
		</td>
		<td>
			10
		</td>
		<td>
			2
		</td>
		<td>
			7
		</td>
		<td>
			16
		</td>
		<td>
			14
		</td>
		<td>
			31
		</td>
	</tr>
</table>

### Building:

- Windows\
    Use solution files.

- Linux
    ```
    Requirements:
        - Git
        - C++17 compiler
        - CMake >= 3.16
    ```
    ```
    git clone https://github.com/Asd-g/AviSynth-tcolormask && \
    cd AviSynth-tcolormask && \
    mkdir build && \
    cd build && \
    cmake .. && \
    make -j$(nproc) && \
    sudo make install
    ```

### License:
This project is licensed under the [MIT license][mit_license]. Binaries are [GPL v2][gpl_v2].

[old_script]: http://pastebin.com/vAa8fyjp
[comparison]: http://screenshotcomparison.com/comparison/27079
[mit_license]: http://opensource.org/licenses/MIT
[gpl_v2]: http://www.gnu.org/licenses/gpl-2.0.html


CHANGELOG.md
##### 1.2.2:
    `grayscale` and `onlyY` cannot be true in the same time.
    Added AVX512 code.

##### 1.2.1:
    Registered as MT_NICE_FILTER. (tp7)
    Disabled internal MT by default. (tp7)
    Fixed YV12 processing. (tp7)
    Added support for frame properties passthrough. (Asd-g)
    Fixed the right side when width is not mod16. (Asd-g)
    Non-mod2 chroma height for YV12 and mt=true are not allowed. (Asd-g)
    Added C and AVX2 code. (Asd-g)
    Added `opt` parameter. (Asd-g)
    Added support for 16-bit clips. (Asd-g)
    Added `onlyY` parameter. (Asd-g)
    Added binary version. (Asd-g)

##### 1.2:
    x64 build added. (tp7)

##### 1.1:
    This plugin now only works with AviSynth 2.6 alpha 4. If you need support for older versions, download v1.0.1.
    Also, YV16 colorspace is now supported. (tp7)

##### 1.0.1:
    Initial release. (tp7)


tcolormask/tcolormask.h
#pragma once

#include <vector>

#include "avisynth.h"

template <typename T>
struct YUVPixel
{
    T Y;
    T U;
    T V;
};

template <typename T, bool grayscale, bool mt>
class TColorMask : public GenericVideoFilter
{
public:
    TColorMask(PClip child, std::vector<uint64_t> colors, int tolerance, bool bt601, int lutthr, bool onlyY, int opt, IScriptEnvironment* env);
    PVideoFrame __stdcall GetFrame(int n, IScriptEnvironment* env) override;

    int __stdcall SetCacheHints(int cachehints, int frame_range) override
    {
        return cachehints == CACHE_GET_MTMODE ? MT_NICE_FILTER : 0;
    }

private:
    void buildLuts() noexcept;

    template <bool cs, bool border>
    void process(T* __restrict dstY_ptr, const T* srcY_ptr, const T* srcV_ptr, const T* srcU_ptr, int dst_pitch_y, int src_pitch_y, int src_pitch_uv, int width, int height) noexcept;
    void (TColorMask::* proc)(T* __restrict dstY_ptr, const T* srcY_ptr, const T* srcV_ptr, const T* srcU_ptr, int dst_pitch_y, int src_pitch_y, int src_pitch_uv, int width, int height) noexcept;

    void(*proc_lut)(T* __restrict pDstY, const T* pSrcY, const T* pSrcV, const T* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, std::vector<T>& lutY, std::vector<T>& lutU, std::vector<T>& lutV) noexcept;
    void(*p_)(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<T>>& colors, int tolerance, int halftolerance) noexcept;

    std::vector<YUVPixel<T>> colors_;
    int tolerance_;
    int subsamplingY_;
    int subsamplingX_;

    std::vector<T> lut_y;
    std::vector<T> lut_u;
    std::vector<T> lut_v;

    bool v8;

    VideoInfo vi1;
};

template <typename T, int subsamplingX, int subsamplingY>
void processSse2(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<T>>& colors, int tolerance, int halftolerance) noexcept;
template <typename T, int subsamplingX, int subsamplingY>
void processAvx2(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<T>>& colors, int tolerance, int halftolerance) noexcept;
template <typename T, int subsamplingX, int subsamplingY>
void processAvx512(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<T>>& colors, int tolerance, int halftolerance) noexcept;


tcolormask/tcolormask_avx2.cpp
#include "tcolormask.h"
#include "VCL2/vectorclass.h"

template <typename T, int subsamplingX, int subsamplingY>
void processAvx2(void* __restrict pDstY_, const void* pSrcY_, const void* pSrcV_, const void* pSrcU_, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<T>>& colors, int tolerance, int halftolerance) noexcept
{
    const T* pSrcY = reinterpret_cast<const T*>(pSrcY_);
    const T* pSrcV = reinterpret_cast<const T*>(pSrcV_);
    const T* pSrcU = reinterpret_cast<const T*>(pSrcU_);
    T* __restrict pDstY = reinterpret_cast<T*>(pDstY_);

    for (int y = 0; y < height; ++y)
    {
        if constexpr (std::is_same_v<T, uint8_t>)
        {
            for (int x = 0; x < width; x += 32)
            {
                Vec32uc result_y = zero_si256();
                Vec32uc result_u = zero_si256();
                Vec32uc result_v = zero_si256();

                const auto srcY_v = Vec32uc().load(pSrcY + x);
                Vec32uc srcU_v, srcV_v;

                if constexpr (subsamplingX == 2)
                {
                    srcU_v = Vec32uc().load(pSrcU + x / subsamplingX);
                    srcU_v = blend32<0, 32, 1, 33, 2, 34, 3, 35, 4, 36, 5, 37, 6, 38, 7, 39, 8, 40, 9, 41, 10, 42, 11, 43, 12, 44, 13, 45, 14, 46, 15, 47>(srcU_v, srcU_v);
                    srcV_v = Vec32uc().load(pSrcV + x / subsamplingX);
                    srcV_v = blend32<0, 32, 1, 33, 2, 34, 3, 35, 4, 36, 5, 37, 6, 38, 7, 39, 8, 40, 9, 41, 10, 42, 11, 43, 12, 44, 13, 45, 14, 46, 15, 47>(srcV_v, srcV_v);
                }
                else
                {
                    srcU_v = Vec32uc().load(pSrcU + x);
                    srcV_v = Vec32uc().load(pSrcV + x);
                }

                for (auto& color : colors)
                {
                    const auto colorVector_y = Vec32uc(color.Y);
                    const auto colorVector_u = Vec32uc(color.U);
                    const auto colorVector_v = Vec32uc(color.V);
                    /* absolute difference */
                    const auto maximum_y = max(srcY_v, colorVector_y);
                    const auto maximum_u = max(srcU_v, colorVector_u);
                    const auto maximum_v = max(srcV_v, colorVector_v);

                    const auto minimum_y = min(srcY_v, colorVector_y);
                    const auto minimum_u = min(srcU_v, colorVector_u);
                    const auto minimum_v = min(srcV_v, colorVector_v);

                    const auto diff_y = maximum_y - minimum_y;
                    const auto diff_u = maximum_u - minimum_u;
                    const auto diff_v = maximum_v - minimum_v;
                    /* comparing to tolerance */
                    const auto diff_tolerance_min_y = max(diff_y, Vec32uc(tolerance));
                    const auto diff_tolerance_min_u = max(diff_u, Vec32uc(halftolerance));
                    const auto diff_tolerance_min_v = max(diff_v, Vec32uc(halftolerance));

                    const auto passed_y = select(diff_y >= diff_tolerance_min_y, zero_si256(), Vec32uc(255));
                    const auto passed_u = select(diff_u >= diff_tolerance_min_u, zero_si256(), Vec32uc(255));
                    const auto passed_v = select(diff_v >= diff_tolerance_min_v, zero_si256(), Vec32uc(255));

                    result_y = result_y | passed_y;
                    result_u = result_u | passed_u;
                    result_v = result_v | passed_v;
                }

                result_y = result_y & result_u;
                result_y = result_y & result_v;
                result_y.store(pDstY + x);
            }
        }
        else
        {
            for (int x = 0; x < width; x += 16)
            {
                Vec16us result_y = zero_si256();
                Vec16us result_u = zero_si256();
                Vec16us result_v = zero_si256();

                const auto srcY_v = Vec16us().load(pSrcY + x);
                Vec16us srcU_v, srcV_v;

                if constexpr (subsamplingX == 2)
                {
                    srcU_v = Vec16us().load(pSrcU + x / subsamplingX);
                    srcU_v = blend16<0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23>(srcU_v, srcU_v);
                    srcV_v = Vec16us().load(pSrcV + x / subsamplingX);
                    srcV_v = blend16<0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23>(srcV_v, srcV_v);
                }
                else
                {
                    srcU_v = Vec16us().load(pSrcU + x);
                    srcV_v = Vec16us().load(pSrcV + x);
                }

                for (auto& color : colors)
                {
                    const auto colorVector_y = Vec16us(color.Y);
                    const auto colorVector_u = Vec16us(color.U);
                    const auto colorVector_v = Vec16us(color.V);
                    /* absolute difference */
                    const auto maximum_y = max(srcY_v, colorVector_y);
                    const auto maximum_u = max(srcU_v, colorVector_u);
                    const auto maximum_v = max(srcV_v, colorVector_v);

                    const auto minimum_y = min(srcY_v, colorVector_y);
                    const auto minimum_u = min(srcU_v, colorVector_u);
                    const auto minimum_v = min(srcV_v, colorVector_v);

                    const auto diff_y = (maximum_y - minimum_y);
                    const auto diff_u = (maximum_u - minimum_u);
                    const auto diff_v = (maximum_v - minimum_v);
                    /* comparing to tolerance */
                    const auto diff_tolerance_min_y = max(diff_y, Vec16us(tolerance));
                    const auto diff_tolerance_min_u = max(diff_u, Vec16us(halftolerance));
                    const auto diff_tolerance_min_v = max(diff_v, Vec16us(halftolerance));

                    const auto passed_y = select(diff_y >= diff_tolerance_min_y, zero_si256(), Vec16us(65535));
                    const auto passed_u = select(diff_u >= diff_tolerance_min_u, zero_si256(), Vec16us(65535));
                    const auto passed_v = select(diff_v >= diff_tolerance_min_v, zero_si256(), Vec16us(65535));

                    result_y = result_y | passed_y;
                    result_u = result_u | passed_u;
                    result_v = result_v | passed_v;
                }

                result_y = result_y & result_u;
                result_y = result_y & result_v;
                result_y.store(pDstY + x);
            }
        }

        pSrcY += srcPitchY;

        if (y % subsamplingY == (subsamplingY - 1))
        {
            pSrcU += srcPitchUV;
            pSrcV += srcPitchUV;
        }

        pDstY += dstPitchY;
    }
}

template void processAvx2<uint8_t, 1, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint8_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx2<uint8_t, 2, 2>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint8_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx2<uint8_t, 2, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint8_t>>& colors,
    int tolerance, int halftolerance) noexcept;

template void processAvx2<uint16_t, 1, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint16_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx2<uint16_t, 2, 2>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint16_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx2<uint16_t, 2, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint16_t>>& colors,
    int tolerance, int halftolerance) noexcept;


tcolormask/tcolormask_avx512.cpp
#include "tcolormask.h"
#include "VCL2/vectorclass.h"

template <typename T, int subsamplingX, int subsamplingY>
void processAvx512(void* __restrict pDstY_, const void* pSrcY_, const void* pSrcV_, const void* pSrcU_, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<T>>& colors, int tolerance, int halftolerance) noexcept
{
    const T* pSrcY = reinterpret_cast<const T*>(pSrcY_);
    const T* pSrcV = reinterpret_cast<const T*>(pSrcV_);
    const T* pSrcU = reinterpret_cast<const T*>(pSrcU_);
    T* __restrict pDstY = reinterpret_cast<T*>(pDstY_);

    for (int y = 0; y < height; ++y)
    {
        if constexpr (std::is_same_v<T, uint8_t>)
        {
            for (int x = 0; x < width; x += 64)
            {
                Vec64uc result_y = zero_si512();
                Vec64uc result_u = zero_si512();
                Vec64uc result_v = zero_si512();

                const auto srcY_v = Vec64uc().load(pSrcY + x);
                Vec64uc srcU_v, srcV_v;

                if constexpr (subsamplingX == 2)
                {
                    srcU_v = Vec64uc().load(pSrcU + x / subsamplingX);
                    srcU_v = blend64<0, 64, 1, 65, 2, 66, 3, 67, 4, 68, 5, 69, 6, 70, 7, 71, 8, 72, 9, 73, 10, 74, 11, 75, 12, 76, 13, 77, 14, 78, 15, 79,
                        16, 80, 17, 81, 18, 82, 19, 83, 20, 84, 21, 85, 22, 86, 23, 87, 24, 88, 25, 89, 26, 90, 27, 91, 28, 92, 29, 93, 30, 94, 31, 95>(srcU_v, srcU_v);
                    srcV_v = Vec64uc().load(pSrcV + x / subsamplingX);
                    srcV_v = blend64<0, 64, 1, 65, 2, 66, 3, 67, 4, 68, 5, 69, 6, 70, 7, 71, 8, 72, 9, 73, 10, 74, 11, 75, 12, 76, 13, 77, 14, 78, 15, 79,
                        16, 80, 17, 81, 18, 82, 19, 83, 20, 84, 21, 85, 22, 86, 23, 87, 24, 88, 25, 89, 26, 90, 27, 91, 28, 92, 29, 93, 30, 94, 31, 95>(srcV_v, srcV_v);
                }
                else
                {
                    srcU_v = Vec64uc().load(pSrcU + x);
                    srcV_v = Vec64uc().load(pSrcV + x);
                }

                for (auto& color : colors)
                {
                    const auto colorVector_y = Vec64uc(color.Y);
                    const auto colorVector_u = Vec64uc(color.U);
                    const auto colorVector_v = Vec64uc(color.V);
                    /* absolute difference */
                    const auto maximum_y = max(srcY_v, colorVector_y);
                    const auto maximum_u = max(srcU_v, colorVector_u);
                    const auto maximum_v = max(srcV_v, colorVector_v);

                    const auto minimum_y = min(srcY_v, colorVector_y);
                    const auto minimum_u = min(srcU_v, colorVector_u);
                    const auto minimum_v = min(srcV_v, colorVector_v);

                    const auto diff_y = maximum_y - minimum_y;
                    const auto diff_u = maximum_u - minimum_u;
                    const auto diff_v = maximum_v - minimum_v;
                    /* comparing to tolerance */
                    const auto diff_tolerance_min_y = max(diff_y, Vec64uc(tolerance));
                    const auto diff_tolerance_min_u = max(diff_u, Vec64uc(halftolerance));
                    const auto diff_tolerance_min_v = max(diff_v, Vec64uc(halftolerance));

                    const auto passed_y = select(diff_y >= diff_tolerance_min_y, zero_si512(), Vec64uc(255));
                    const auto passed_u = select(diff_u >= diff_tolerance_min_u, zero_si512(), Vec64uc(255));
                    const auto passed_v = select(diff_v >= diff_tolerance_min_v, zero_si512(), Vec64uc(255));

                    result_y = result_y | passed_y;
                    result_u = result_u | passed_u;
                    result_v = result_v | passed_v;
                }

                result_y = result_y & result_u;
                result_y = result_y & result_v;
                result_y.store(pDstY + x);
            }
        }
        else
        {
            for (int x = 0; x < width; x += 32)
            {
                Vec32us result_y = zero_si512();
                Vec32us result_u = zero_si512();
                Vec32us result_v = zero_si512();

                const auto srcY_v = Vec32us().load(pSrcY + x);
                Vec32us srcU_v, srcV_v;

                if constexpr (subsamplingX == 2)
                {
                    srcU_v = Vec32us().load(pSrcU + x / subsamplingX);
                    srcU_v = blend32<0, 32, 1, 33, 2, 34, 3, 35, 4, 36, 5, 37, 6, 38, 7, 39, 8, 40, 9, 41, 10, 42, 11, 43, 12, 44, 13, 45, 14, 46, 15, 47>(srcU_v, srcU_v);
                    srcV_v = Vec32us().load(pSrcV + x / subsamplingX);
                    srcV_v = blend32<0, 32, 1, 33, 2, 34, 3, 35, 4, 36, 5, 37, 6, 38, 7, 39, 8, 40, 9, 41, 10, 42, 11, 43, 12, 44, 13, 45, 14, 46, 15, 47>(srcV_v, srcV_v);
                }
                else
                {
                    srcU_v = Vec32us().load(pSrcU + x);
                    srcV_v = Vec32us().load(pSrcV + x);
                }

                for (auto& color : colors)
                {
                    const auto colorVector_y = Vec32us(color.Y);
                    const auto colorVector_u = Vec32us(color.U);
                    const auto colorVector_v = Vec32us(color.V);
                    /* absolute difference */
                    const auto maximum_y = max(srcY_v, colorVector_y);
                    const auto maximum_u = max(srcU_v, colorVector_u);
                    const auto maximum_v = max(srcV_v, colorVector_v);

                    const auto minimum_y = min(srcY_v, colorVector_y);
                    const auto minimum_u = min(srcU_v, colorVector_u);
                    const auto minimum_v = min(srcV_v, colorVector_v);

                    const auto diff_y = (maximum_y - minimum_y);
                    const auto diff_u = (maximum_u - minimum_u);
                    const auto diff_v = (maximum_v - minimum_v);
                    /* comparing to tolerance */
                    const auto diff_tolerance_min_y = max(diff_y, Vec32us(tolerance));
                    const auto diff_tolerance_min_u = max(diff_u, Vec32us(halftolerance));
                    const auto diff_tolerance_min_v = max(diff_v, Vec32us(halftolerance));

                    const auto passed_y = select(diff_y >= diff_tolerance_min_y, zero_si512(), Vec32us(65535));
                    const auto passed_u = select(diff_u >= diff_tolerance_min_u, zero_si512(), Vec32us(65535));
                    const auto passed_v = select(diff_v >= diff_tolerance_min_v, zero_si512(), Vec32us(65535));

                    result_y = result_y | passed_y;
                    result_u = result_u | passed_u;
                    result_v = result_v | passed_v;
                }

                result_y = result_y & result_u;
                result_y = result_y & result_v;
                result_y.store(pDstY + x);
            }
        }

        pSrcY += srcPitchY;

        if (y % subsamplingY == (subsamplingY - 1))
        {
            pSrcU += srcPitchUV;
            pSrcV += srcPitchUV;
        }

        pDstY += dstPitchY;
    }
}

template void processAvx512<uint8_t, 1, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint8_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx512<uint8_t, 2, 2>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint8_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx512<uint8_t, 2, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint8_t>>& colors,
    int tolerance, int halftolerance) noexcept;

template void processAvx512<uint16_t, 1, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint16_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx512<uint16_t, 2, 2>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint16_t>>& colors,
    int tolerance, int halftolerance) noexcept;
template void processAvx512<uint16_t, 2, 1>(void* __restrict pDstY, const void* pSrcY, const void* pSrcV, const void* pSrcU, int dstPitchY, int srcPitchY, int srcPitchUV, int width, int height, const std::vector<YUVPixel<uint16_t>>& colors,
    int tolerance, int halftolerance) noexcept;
